### PR TITLE
[ISSUE-668] Update ac before lvg deletion

### LINKED ISSUE #668: No need to remove AC on DeleteVolume request
**Describe the bug**
No need to remove AC on DeleteVolume request. 

**Environment :**
 - Version: 1.1.0

**To Reproduce**

1. Deploy statefulset with LVG. 
2. Delete statefulset.
3. Remove volume with lvg.

**Expected behavior**
After volume with lvg removing, corresponding AC was deleted, and recreated.

**Additional context**
Changes are needed [here](https://github.com/dell/csi-baremetal/blob/49cd1467eff3c6a1ce7dc272810ded66b8432f2e/pkg/common/volume_operations.go#L639-L645).

### PR DESCRIPTION
**Pull request by yurkov-anton**
### Resolves #668

Add convert AC to Drive before LVG deletion.
Refactoring LVG deletion.
Add waiting LVG deletion before move drive to REMOVED.

## PR checklist
- [ ] Add link to the issue
- [ ] Choose Project
- [ ] Choose PR label
- [ ] New unit tests added
- [ ] Modified code has meaningful comments
- [ ] All TODOs are linked with the issues
- [ ] All comments are resolved

## Testing
Test Drive Replacement, with LVG volumes

### CODE REVIEW

File: devkit/devkit [hunk 1]
```diff
@@ -1,4 +1,4 @@
-#!/bin/bash
+#!/bin/bash -x
```
Comment (safronovD) [inline devkit/devkit:1]: Is it the related changes?

File: pkg/common/volume_operations.go [hunk 2]
```diff
@@ -426,53 +426,75 @@ func (vo *VolumeOperationsImpl) UpdateCRsAfterVolumeDeletion(ctx context.Context
...
-			ll.Errorf("Unable to remove volume reference from LogicalVolumeGroup %s: %v", volumeCR.Spec.Location, err)
-		}
 	}
 
-	// if LogicalVolumeGroup wasn't deleted and health of volume is GOOD increase AC size
-	// We don't increase AC size for unhealthy volume to avoid new allocations on top of unhealthy drive/lvg
-	if !isDeleted && volumeCR.Spec.Health == apiV1.HealthGood {
+	if volumeCR.Spec.Health == apiV1.HealthGood {
 		// Increase size of AC using volume size
 		acCR.Spec.Size += volumeCR.Spec.Size
```
Comment (safronovD) [inline pkg/common/volume_operations.go:450]: If we delete LVG Volume, do we update AC 2 times (here and in `updateLVGAfterVolumeDeletion`)? Could we avoid it?

File: pkg/common/volume_operations.go [hunk 3]
```diff
@@ -426,53 +426,75 @@ func (vo *VolumeOperationsImpl) UpdateCRsAfterVolumeDeletion(ctx context.Context
...
 
+func (vo *VolumeOperationsImpl) updateLVGAfterVolumeDeletion(ctx context.Context, volumeCR *volumecrd.Volume,
+	driveCR *drivecrd.Drive, acCR *accrd.AvailableCapacity) error {
+	ll := vo.log.WithFields(logrus.Fields{
+		"method":   "updateLVGAfterVolumeDeletion",
+		"volumeID": volumeCR.Name,
+	})
+
+	lvg := lvgcrd.LogicalVolumeGroup{}
+	if err := vo.k8sClient.ReadCR(ctx, volumeCR.Spec.Location, "", &lvg); err != nil {
```
Comment (safronovD) [inline pkg/common/volume_operations.go:466]: We have already read LVGCR in `vo.crHelper.GetACByLocation`. Could we use that one?

File: pkg/common/volume_operations.go [hunk 4]
```diff
@@ -426,53 +426,75 @@ func (vo *VolumeOperationsImpl) UpdateCRsAfterVolumeDeletion(ctx context.Context
...
+	// If volume not last, remove from lvg ref list, If volume is last need convert AC to Drive and remove LVG
+	if len(lvg.Spec.VolumeRefs) > 1 {
+		lvg.Spec.VolumeRefs = util.RemoveString(lvg.Spec.VolumeRefs, volumeCR.Name)
+		ll.Debugf("Remove volume ref %s from LVG %s", volumeCR.Name, lvg.Name)
+		if err := vo.k8sClient.UpdateCR(ctx, &lvg); err != nil {
+			ll.Errorf("Unable to update LVG %s CR: %v", lvg.Name, err)
+			return err
+		}
+	} else {
+		ll.Debugf("Delete LVG %v", lvg.Name)
```
Comment (safronovD) [inline pkg/common/volume_operations.go:480]: In my opinion we should update AC at first and then delete LVG. As I see in drive_controller changes, we check LVG before AC deletion try, so it my couse to:
1. controller: delete LVG
2. drive_controller: check LVG 
3. drive_controller: try to delete AC and fail
4. controller: update AC

File: devkit/devkit (hunk 1 above)
Comment (yurkov-anton) [reply devkit/devkit:1]: no, it's extra

File: pkg/common/volume_operations.go (hunk 2 above)
Comment (yurkov-anton) [reply pkg/common/volume_operations.go:450]: yes, AC will be updated 2 times. But for do one update need to do more complex solve. And I'm not sure that need to unite this updates, it's two unrelated updates: 
1. add size due volume deletion
2. convert AC due LVG deletion

File: pkg/common/volume_operations.go (hunk 3 above)
Comment (yurkov-anton) [reply pkg/common/volume_operations.go:466]: sure

File: pkg/common/volume_operations.go (hunk 4 above)
Comment (yurkov-anton) [reply pkg/common/volume_operations.go:480]: yes, it will be changed

**Review by safronovD — APPROVED** [review]

**Review by rkiselenko — APPROVED** [review]

**Review by mishoyama — APPROVED** [review]

**Review by misanthropicat — APPROVED** [review]

**Review by rkiselenko — APPROVED** [review]

**Pull request merged**